hypothesis-python/src/hypothesis/extra/gufunc.py

```diff
@@ -0,0 +1,451 @@
+# This module uses the numpy parser of the Generalized Universal Function API
+# signatures `_parse_gufunc_signature`, which is only available in
+# numpy>=1.12.0 and therefore requires a bump in the requirements for
+# hypothesis.
+from __future__ import absolute_import, division, print_function
+
+from collections import defaultdict
+
+import numpy as np
+import numpy.lib.function_base as npfb
+
+from hypothesis.errors import InvalidArgument
+from hypothesis.extra.numpy import arrays, order_check
+from hypothesis.internal.validation import (
+    check_valid_bound,
+    check_valid_interval,
+)
+from hypothesis.searchstrategy import SearchStrategy
+from hypothesis.strategies import (
+    builds,
+    composite,
+    fixed_dictionaries,
+    integers,
+    just,
+    tuples,
+)
+
+# Should not ever need to broadcast beyond this, but should be able to set it
+# as high as 32 before breaking assumptions in numpy.
+GLOBAL_DIMS_MAX = 12
+
+# Key used in min_side and max_side to indicate min/max on broadcasted dims,
+# using ``object()`` trick to create unique sentinel.
+BCAST_DIM = object()
+# Value used in default dict for max side if variable not specified
+DEFAULT_MAX_SIDE = 5
+
+# This uses "private" function of numpy, but it does the job. It throws a
+# pretty readable exception for invalid input, so we don't need to add anything
+# there. We should also check signature.isascii() since there are lot of weird
+# corner cases with unicode parsing, but isascii() restricts us to Py >=3.7.
```

Comment:
I meant rather than if we require even Python3 strings to stay within the printable ASCII characters we'll be fine, as non-latin digits such as `'١'` have codepoints > 127.

Reply:
It might be easier to support all unicode except the weird digits, because then we can generate the test cases for `gufunc_arg_shapes` using
`from_regex(npfb._SIGNATURE).filter(no_weird_digits)`
in both Python2 and Python3.

Reply:
I think even `from_regex(npfb._SIGNATURE.replace("\d", "[0-9]"))` would work.

Reply:
The question is though, do we want `gufunc` to accept signatures with higher unicode characters (except the non-latin digits)?  Or should we restrict signatures to ASCII only?  The tests are certainly more exhaustive with ASCII only, but maybe this goes against the spirit of python 3 unicode everywhere.

Reply:
I'd make our lives easier by restricting it to ASCII until someone actually asks us for unicode support.

Reply:
ok, so then we should change it to this?
```python
def parse_gufunc_signature(signature):
    if any(ord(cc) > 127 for cc in signature):
        raise InvalidArgument("Cannot handle non-ASCII digit characters in signature=%r" % (signature,))
    return npfb._parse_gufunc_signature(signature)
```

at least until we can require python 3.7 in the future and use `isascii`.

Reply:
Yep, though replace "non-ASCII digit characters" with "non-ASCII characters".